### PR TITLE
fix: handle insufficient balance on receiving chain (for protocol fees)

### LINKED ISSUE #4350: Lifi trades with input of Eth on Optimism not working
### Overview

when trying to trade eth on optimism to bnb on bnb the trade is failing on native and on KK the trade modal keeps resetting on confirm and trade and doesn't complete

### References and additional details

1) Log into app.shapeshift and connect to native wallet
2) Navigate to trade page and enter an Eth on Optimism to anything trade
3) Preview trade and Confirm and trade
4) Note that the trade modal continues to spin as pending even tho a tx shows up in recent transactions as a failed send.
(On KK, the trade modal will confirm and trade will disappear for a bit then become available again to select. Prompts on kk itself are slow and don't execute the trade)
image
https://optimistic.etherscan.io/tx/0x12c4d770a1b666f4b840ef9c03325697116cba4d8f4c6a2c0f31a27292b07831
https://user-images.githubusercontent.com/100157997/234437329-3fb99d64-4d32-4fcd-b47a-bb8d9b7d1433.mp4

 

### Acceptance Criteria

Should be able to complete a trade from Eth on Optimism 

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Ownership

- [X] If my bounty needs engineering or needs product I have added the respective labels on the right
- [X] As the sponsor of this bounty I will review the changes in a preview environment (ops/product) or review the PR (engineering)

### Estimated effort

_No response_

### Sponsor / Stakeholder

_No response_

### Bounty Hunters

- [Join our discord](https://discord.gg/shapeshift)
- Include an expected timeline for you to complete work in the work plan when you apply for this bounty!
- Please refer to [this link](https://shapeshiftdao.gitbook.io/getting-started/how-to-contribute/bounties#resources-for-new-contributors-and-bounty-hunters) for some basic info
- Please ***do not*** start work on this issue until you are approved in Gitcoin.

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Fixes dead click when attempting trades with insufficient balance for protocol fees on receiving chain.

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #4350

## Risk

low risk of trades breaking.

## Testing

- check trades are unaffected
- check fee amounts are still correct
- check general warnings about invalid trade states are still correct
- check 2nd protocol fee is correct for `any -> matic on polygon` trades
- check insufficient balance warning is correct for protocol fees on receiving chain

### Engineering

Note that this is as pragmatic a solution as I could push without boiling the ocean. When we support multi-hop we'll need to remove the concept of `buyAssetTradeFeeUsd` and `sellAssetTradeFeeUsd` and instead return an array of fees since there will be intermediary fees in other tokens. This will be a fairly heavy lift in thor swapper but is pretty straight forwards otherwise.

### Operations

Please check #4350 to verify the issue is resolved.

## Screenshots (if applicable)
case with 2 protocol fees:
image

sometimes there is only protocol fee:
image

### CODE REVIEW

File: src/components/Trade/TradeConfirm/ReceiveSummary.tsx [hunk 1]
```diff
@@ -7,35 +7,47 @@ import {
...
+import { fromBaseUnit } from 'lib/math'
+import type { AmountDisplayMeta } from 'lib/swapper/api'
 import { SwapperName } from 'lib/swapper/api'
 
 type ReceiveSummaryProps = {
   isLoading?: boolean
   symbol: string
-  amount: string
-  intermediaryTransactionOutputs?: { amount: string; symbol: string; chainName?: string }[]
+  amountCryptoPrecision: string
```
Comment (gomesalexandre) [inline src/components/Trade/TradeConfirm/ReceiveSummary.tsx:26]: tyvm

File: src/components/Trade/TradeConfirm/ReceiveSummary.tsx [hunk 2]
```diff
@@ -7,35 +7,47 @@ import {
...
-  protocolFee?: string
+  amountBeforeFeesCryptoPrecision?: string
+  protocolFees?: AmountDisplayMeta[]
   shapeShiftFee?: string
   slippage: string
   swapperName: string
 } & RowProps
 
+const parseAmountDisplayMeta =
+  (chainAdapterManager: ChainAdapterManager) =>
```
Comment (gomesalexandre) [inline src/components/Trade/TradeConfirm/ReceiveSummary.tsx:37]: Any reason why we inject this? This is a singleton, thus should have no perf. implications to get from within this method

File: src/lib/swapper/swappers/LifiSwapper/utils/getIntermediaryTransactionOutputs/getIntermediaryTransactionOutputs.ts [hunk 3]
```diff
@@ -33,9 +33,9 @@ export const getIntermediaryTransactionOutputs = (
       }
 
       return {
-        asset,
-        buyAmountCryptoBaseUnit: step.estimate.toAmountMin,
+        ...asset,
+        amountCryptoBaseUnit: step.estimate.toAmountMin,
       }
     })
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/LifiSwapper/utils/getIntermediaryTransactionOutputs/getIntermediaryTransactionOutputs.ts:39]: `Object.assign(asset, { amountCryptoBaseUnit }` if you feel like it

File: src/lib/swapper/swappers/LifiSwapper/utils/getIntermediaryTransactionOutputs/getIntermediaryTransactionOutputs.ts [hunk 4]
```diff
@@ -33,9 +33,9 @@ export const getIntermediaryTransactionOutputs = (
...
-    .filter((item): item is IntermediaryTransactionOutput => item !== undefined)
+    .filter((item): item is AmountDisplayMeta => item !== undefined)
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/LifiSwapper/utils/getIntermediaryTransactionOutputs/getIntermediaryTransactionOutputs.ts:40]: `isSome` ?

File: src/state/zustand/swapperStore/amountSelectors.ts [hunk 5]
```diff
@@ -199,8 +199,39 @@ export const selectBuyAssetTradeFeeBuyAssetBaseUnit = createSelector(
...
+    if (buyAssetTradeFeeCryptoBaseUnit) {
+      result.push({
+        ...buyAsset,
+        amountCryptoBaseUnit: buyAssetTradeFeeCryptoBaseUnit,
+      })
+    }
+
+    return result
+  },
+)
```
Comment (gomesalexandre) [inline src/state/zustand/swapperStore/amountSelectors.ts:231]: Ditto Object.assign if you feel like it

File: src/lib/swapper/swappers/LifiSwapper/utils/getIntermediaryTransactionOutputs/getIntermediaryTransactionOutputs.test.ts [hunk 6]
```diff
@@ -26,9 +26,9 @@ describe('getIntermediaryTransactionOutputs', () => {
     expect(usdcOnEthereum).not.toBe(undefined)
     expect(usdcOnOptimism).not.toBe(undefined)
 
-    const expectation: IntermediaryTransactionOutput[] = [
-      { asset: usdcOnEthereum, buyAmountCryptoBaseUnit: '120527596' },
-      { asset: usdcOnOptimism, buyAmountCryptoBaseUnit: '120328249' },
+    const expectation: AmountDisplayMeta[] = [
+      { ...usdcOnEthereum, amountCryptoBaseUnit: '120527596' },
+      { ...usdcOnOptimism, amountCryptoBaseUnit: '120328249' },
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/LifiSwapper/utils/getIntermediaryTransactionOutputs/getIntermediaryTransactionOutputs.test.ts:31]: Ditto don't spread and would prefer to keep the partial asset as a separate property

File: src/state/zustand/swapperStore/amountSelectors.ts [hunk 7]
```diff
@@ -199,8 +199,39 @@ export const selectBuyAssetTradeFeeBuyAssetBaseUnit = createSelector(
   },
 )
 
+export const selectProtocolFees = createDeepEqualOutputSelector(
```
Comment (0xApotheosis) [inline src/state/zustand/swapperStore/amountSelectors.ts:202]: This is handy - we'll be able to update this selector as we move to the new fee array, and the consumers will be non-the-wiser.

File: src/components/Trade/TradeConfirm/ReceiveSummary.tsx [hunk 8]
```diff
@@ -52,7 +61,33 @@ export const ReceiveSummary: FC<ReceiveSummaryProps> = ({
...
+    [protocolFees],
+  )
+
+  const intermediaryTransactionOutputsParsed = useMemo(
+    () =>
+      intermediaryTransactionOutputs
+        ?.filter(({ amountCryptoBaseUnit }) => bnOrZero(amountCryptoBaseUnit).gt(0))
+        .map(parseAmountDisplayMeta),
+    [intermediaryTransactionOutputs],
+  )
```
Comment (0xApotheosis) [inline src/components/Trade/TradeConfirm/ReceiveSummary.tsx:80]: Could add a `parseAmountDisplayMeta` for these.

**Review by 0xApotheosis — APPROVED** [review]
LGTM.

Gomes' comments have been addressed, code is sane, did not observe any regressions.

Some non-blocking comments we can tackle later if need-be.

**Pull request merged**